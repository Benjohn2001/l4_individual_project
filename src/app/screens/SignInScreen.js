import React, { useState, useRef } from "react";
import {
  View,
  Image,
  TextInput,
  Text,
  TouchableOpacity,
  SafeAreaView,
  Alert,
} from "react-native";
import {
  signInWithEmailAndPassword,
  sendPasswordResetEmail,
} from "firebase/auth";
import Modal from "react-native-modal";
import AppButtonPurple from "../components/AppButtonPurple";
import TwoButtonsSide from "../components/TwoButtonsSide";
import COLOURS from "../assets/colours";
import { auth } from "../../firebase";

function SignInScreen({ navigation }) {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [emailReset, setEmailReset] = useState("");
  const [isModalVisible, setModalVisible] = React.useState(false);

  const signinUser = () => {
    if (email === "" || password === "") {
      Alert.alert("Incomplete form", "Please fill out the full form");
    } else {
      signInWithEmailAndPassword(auth, email, password)
        .then(() => {
          setEmail("");
          setPassword("");
          navigation.push("HomeScreen");
        })
        .catch((error) => {
          if (error.code === "auth/user-not-found") {
            Alert.alert("User not found", "Please retry entering your email");
          } else if (error.code === "auth/wrong-password") {
            Alert.alert(
              "Wrong password",
              "Please retry entering your password"
            );
          } else if (error.code === "auth/invalid-email") {
            Alert.alert("Invalid email", "Please retry entering your email");
          }
        });
    }
  };

  const passwordRef = useRef();

  return (
    <SafeAreaView className="flex-1 items-center pt-20 bg-primaryPurple">
      <Image
        className="w-64 h-64"
        source={{ uri: Image.resolveAssetSource(require("../assets/clock.png")).uri }}
      />
      <View className="pt-10">
        <TextInput
          className="bg-secondaryPurple my-5 w-80 h-12 rounded-md"
          placeholder="Email"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          value={email}
          onChangeText={(val) => setEmail(val)}
          returnKeyType="next"
          onSubmitEditing={() => {
            passwordRef.current.focus();
          }}
          blurOnSubmit={false}
          autoCapitalize="none"
          testID="unameT"
        />
        <TextInput
          className="bg-secondaryPurple mt-5 w-80 h-12 rounded-md"
          placeholder="Password"
          underlineColorAndroid="transparent"
          cursorColor={COLOURS.darkerPurple}
          secureTextEntry
          value={password}
          onChangeText={(val) => setPassword(val)}
          ref={passwordRef}
          autoCapitalize="none"
          testID="passT"
        />
        <TouchableOpacity
          onPress={() => {
            setModalVisible(true);
          }}
        >
          <Text className="mt-2 mb-20 underline text-darkerPurple">
            Forgotten Password
          </Text>
        </TouchableOpacity>
      </View>
      <TwoButtonsSide
        title1="Register"
        onPress1={() => {
          navigation.navigate("RegisterScreen");
        }}
        color1="#E7E7FF"
        icon1="edit"
        fontColor1="#6B4EFF"
        title2="Sign In"
        onPress2={() => {
          signinUser();
        }}
        icon2="arrow-right-circle"
        color2="#6B4EFF"
        fontColor2="white"
      />

      <Modal
        isVisible={isModalVisible}
        onBackdropPress={() => setModalVisible(false)}
        className="items-center"
      >
        <View className=" w-11/12 h-50 items-center py-3 px-3 bg-secondaryPurple rounded-2xl">
          <Text className="text-2xl font-bold  pb-2">Forgotten Password</Text>
          <Text className="text-center text-gray-500 py-4">
            Enter the email of the account to be reset
          </Text>
          <TextInput
            className="bg-white my-5 w-full mx-3 h-12 rounded-md"
            placeholder="Email"
            underlineColorAndroid="transparent"
            cursorColor={COLOURS.darkerPurple}
            value={emailReset}
            onChangeText={(val) => setEmailReset(val)}
            testID="forgotP"
          />
          <AppButtonPurple
            title="Send Email"
            onPress={() => {
              sendPasswordResetEmail(auth, emailReset).then(() => {
                Alert.alert(
                  "Email sent",
                  `Reset email has been sent to ${emailReset}`
                );
                setModalVisible(false);
                setEmailReset("");
              });
            }}
          />
        </View>
      </Modal>
    </SafeAreaView>
  );
}

export default SignInScreen;
