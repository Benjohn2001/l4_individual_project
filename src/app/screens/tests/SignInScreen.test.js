import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { Alert } from "react-native";
import SignInScreen from "../SignInScreen";

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
  push: jest.fn().mockReturnThis(),
  ref: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

const navigation = {
  navigate: jest.fn(),
  push: jest.fn(),
};

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  auth: jest.fn().mockReturnThis(),
  signInWithEmailAndPassword: () => ({
    then: () => ({
      catch: jest.fn(),
    }),
  }),
  sendPasswordResetEmail: () => ({
    then: jest.fn(),
  }),
  getAuth: () => ({
    signOut: () => ({
      then: () => ({
        catch: jest.fn(),
      }),
    }),
  }),
}));

test("Check SignInScreen login working correctly", () => {
  const { getByTestId } = render(<SignInScreen navigation={navigation} />);

  fireEvent.changeText(getByTestId("unameT"), "abc");
  fireEvent(getByTestId("unameT"), "submitEditing");
  fireEvent.changeText(getByTestId("passT"), "12345678B");
  fireEvent.press(screen.getByText("Sign In"));

  const domTree = renderer.create(<SignInScreen />).toJSON();
  expect(domTree).toMatchSnapshot();
});

test("Check SignInScreen no login data working correctly", () => {
  jest.spyOn(Alert, "alert");
  render(<SignInScreen navigation={navigation} />);

  fireEvent.press(screen.getByText("Sign In"));
  expect(Alert.alert).toHaveBeenCalledTimes(1);
});

test("Check SignInScreen forgotPass working correctly", () => {
  jest.spyOn(Alert, "alert");
  const { getByTestId } = render(<SignInScreen navigation={navigation} />);

  fireEvent.press(screen.getByText("Forgotten Password"));
  fireEvent.changeText(getByTestId("forgotP"), "abc");
  fireEvent.press(screen.getByText("Send Email"));
  expect(Alert.alert).toHaveBeenCalledTimes(1);
});

test("Check SignInScreen nav to register", () => {
  render(<SignInScreen navigation={navigation} />);

  fireEvent.press(screen.getByText("Register"));
  expect(navigation.navigate).toBeCalledWith("RegisterScreen");
});
