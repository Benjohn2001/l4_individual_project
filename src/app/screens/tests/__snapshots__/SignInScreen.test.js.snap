// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check SignInScreen login working correctly 1`] = `
<RCTScrollView
  style={
    Array [
      Object {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "paddingTop": 80,
      },
      Object {
        "backgroundColor": "#C6C4FF",
      },
    ]
  }
>
  <View>
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Array [
          Object {
            "alignItems": "center",
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "uri",
          }
        }
        style={
          Array [
            Object {
              "width": 256,
            },
            Object {
              "height": 256,
            },
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "paddingTop": 40,
            },
          ]
        }
      >
        <TextInput
          autoCapitalize="none"
          blurOnSubmit={false}
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Email"
          returnKeyType="next"
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginBottom": 20,
                "marginTop": 20,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="unameT"
          underlineColorAndroid="transparent"
          value=""
        />
        <TextInput
          autoCapitalize="none"
          cursorColor="#6B4EFF"
          onChangeText={[Function]}
          placeholder="Password"
          secureTextEntry={true}
          style={
            Array [
              Object {
                "backgroundColor": "#E7E7FF",
              },
              Object {
                "marginTop": 20,
              },
              Object {
                "width": 320,
              },
              Object {
                "height": 48,
              },
              Object {
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
              },
            ]
          }
          testID="passT"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "marginTop": 8,
                },
                Object {
                  "marginBottom": 80,
                },
                Object {
                  "textDecorationLine": "underline",
                },
                Object {
                  "color": "#6B4EFF",
                },
              ]
            }
          >
            Forgotten Password
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#E7E7FF",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "height": 48,
              "marginLeft": 6,
              "marginRight": 6,
              "opacity": 1,
              "paddingTop": 12,
              "width": 160,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                Object {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text />
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "marginLeft": 20,
                  },
                  Object {
                    "color": "#6B4EFF",
                  },
                ]
              }
            >
              Register
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#6B4EFF",
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "height": 48,
              "marginLeft": 6,
              "marginRight": 6,
              "opacity": 1,
              "paddingTop": 12,
              "width": 160,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "flexDirection": "row",
                },
                Object {
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text />
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  Object {
                    "textAlign": "center",
                  },
                  Object {
                    "marginLeft": 20,
                  },
                  Object {
                    "color": "white",
                  },
                ]
              }
            >
              Sign In
            </Text>
          </View>
        </View>
      </View>
      <Modal
        animationType="none"
        deviceHeight={null}
        deviceWidth={null}
        hardwareAccelerated={false}
        hideModalContentWhileAnimating={false}
        onBackdropPress={[Function]}
        onModalHide={[Function]}
        onModalWillHide={[Function]}
        onModalWillShow={[Function]}
        onRequestClose={[Function]}
        panResponderThreshold={4}
        scrollHorizontal={false}
        scrollOffset={0}
        scrollOffsetMax={0}
        scrollTo={null}
        statusBarTranslucent={false}
        supportedOrientations={
          Array [
            "portrait",
            "landscape",
          ]
        }
        swipeThreshold={100}
        transparent={true}
        visible={false}
      />
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
